spam/spam_tutorial.py

```diff
@@ -140,234 +140,400 @@
 # * *Third-party models*: using an pre-trained model (usually a model for a different task than the one at hand)
 # * *Distant supervision*: using external knowledge base
 # * *Crowdworker labels*: treating each crowdworker as a black-box function that assigns labels to subsets of the data
+
+# %% [markdown]
+# ### Recommended practice for LF development
+
+# %% [markdown]
+# Typical LF development cycles include multiple iterations of ideation, refining, evaluation, and debugging.
+# A typical cycle consists of the following steps
+#
+# 1. Look at examples to generate ideas for LFs
+# 1. Write an initial version of an LF
+# 1. Estimate its performance by looking at examples in the training set or dev set
+# 1. Refine and debug to improve coverage or accuracy as necessary
+#
+# Labeling function development is a creative process.
+# However, there are a few principles that we highly recommend using when developing labeling functions.
+#
+# * **Use the unlabeled training set for ideation whenever possible.** Reserve labeled examples for evaluation in order to avoid overfitting LFs or models.
+# * **Develop labeling functions in isolation.** Avoid using a "boosting" approach, trying to develop LFs specifically to interact with behavior of existing LFs. If you use the existing LFs to identify uncovered areas of the data, generalize the observed pattern rather than writing an LF to cover those exact examples.
```

Comment:
While a nice principle (that we should probably include), I feel like it's not always true. e.g., the LF that @henryre wrote for Spouses that said "if you have a super spammy word, abstain (since those are already super well covered); otherwise, do this thing based on named entities" and it resulted in a much more useful LF. We definitely don't want to recommend that style of LF writing, especially for beginners, but just wanted to call out that we do break this rule intentionally sometimes.

Reply:
will add a note